lib/bolt/catalog.rb

```diff
@@ -76,7 +76,15 @@ def compile_catalog(request)
       target = request['target']
       plan_vars = shadow_vars('plan', request['plan_vars'], target['facts'])
       target_vars = shadow_vars('target', target['variables'], target['facts'])
-      topscope_vars = target_vars.merge(plan_vars)
+
+      # Merge plan vars with target vars, while maintaining the order of the plan
+      # vars. It's critical that the order of plan vars is not changed, as Puppet
+      # will deserialize the variables in the order they appear. Variables may
+      # contain local references to variables that appear earlier in a plan. If
+      # these variables are moved before the variable they reference, Puppet will
+      # be unable to deserialize the data and raise an error.
+      topscope_vars = target_vars.reject { |k, _v| plan_vars.key?(k) }.merge(plan_vars)
```

Comment:
Only one line of code? What a simple issue!

Reply:
![image](https://user-images.githubusercontent.com/30373575/89233080-b61a9a80-d59d-11ea-81c7-7ba7c5b46db4.png)